app/master/cluster_master.py

```diff
@@ -53,8 +58,36 @@ def __init__(self):
             fs.async_delete(self._master_results_path)
         fs.create_dir(self._master_results_path)
 
+        # Configure heartbeat tracking
+        self._unresponsive_slaves_cleanup_interval = Configuration['unresponsive_slaves_cleanup_interval']
+        self._hb_scheduler = sched.scheduler()
+
         SlavesCollector.register_slaves_metrics_collector(lambda: self.all_slaves_by_id().values())
 
+    def start_heartbeat_tracker_thread(self):
+        self._logger.info('Heartbeat tracker will run every {} seconds'.format(
+            self._unresponsive_slaves_cleanup_interval))
+        SafeThread(target=self._start_heartbeat_tracker, name='HeartbeatTrackerThread', daemon=True).start()
```

Comment:
I'm not sure I agree with this use of `SafeThread` (or with most uses of `SafeThread`). I know @josephharrington explicitly said to use it here, but I disagree. `SafeThread` will kill the whole process if an error bubbles to the top of it. My understanding of the history of SafeThread is that it was introduced in a time before we had alerts in cluster runner. The only way of knowing if an error happened was to kill the whole process. 

But now we have alerts (at least on the master). If an error bubbles up to here, I wouldn't want to kill the whole process, i'd just want to log an error, trigger an alert and keep chugging along. Especially considering that this thread isn't handling anything super duper business critical. It's not as if failures in this thread will make CR unable to run builds or anything like that. Crashing the master seems like an over reaction.

Reply:
To clarify, I agree with Evan that this should definitely not be a SafeThread. The main exception to that that I can think of is if we ever extend this to be a generic periodic task thread using an event loop (for example to support generic periodic tasks like doing git gc on repos, purging old build data, etc.) Even if we do make it a SafeThread in that case, we'd still probably want a try/except around each individual task execution. (Individual tasks should be allowed to fail without killing the entire periodic execution mechanism.)